File: backend/nodejs-typescript/src/middlewares/error.middleware.ts
import { NextFunction, Request, Response } from "express";
import { ZodError } from "zod";
import { ResponseError } from "@/utils/response-error.util";

export default async function errorMiddleware(
  error: Error,
  req: Request,
  res: Response,
  next: NextFunction,
) {
  if (!error) {
    next();
    return;
  }

  req.error = error;

  if (error instanceof ResponseError) {
    res.status(error.status).json({
      errorMsg: error.message,
      errors: error.errors,
    });
    return;
  }

  if (error instanceof ZodError) {
    res.status(400).json({
      errorMsg: "Validation error",
      errors: error.issues,
    });
    return;
  }

  res.status(500).json({
    errorMsg: "Something went wrong. Check logs for detail errors!",
  });
}

File: backend/go/internal/middlewares/error_middleware.go
package middlewares

import (
	"errors"

	"github.com/fathurmdr/backend/go/internal/utils"
	"github.com/gofiber/fiber/v2"
)


func ErrorMiddleware(ctx *fiber.Ctx, err error) error {
	code := fiber.StatusInternalServerError

	var responseError *utils.ResponseError
	if errors.As(err, &responseError) {
		ctx.Status(responseError.Status)

		if(responseError.Errors != nil) {
			return ctx.JSON(fiber.Map{
				"errorMsg": responseError.Message,
				"errors": responseError.Errors,
			})
		}
		
		return ctx.JSON(fiber.Map{
			"errorMsg": responseError.Message,
		})
	}

	var e *fiber.Error
	if errors.As(err, &e) {
			code = e.Code
	}

	return ctx.Status(code).SendString(err.Error())
}